ui/src/util/api.ts

```diff
@@ -0,0 +1,123 @@
+import { getDefaultFetchInit } from '@splunk/splunk-utils/fetch';
+import { app } from '@splunk/splunk-utils/config';
+import { createRESTURL } from '@splunk/splunk-utils/url';
+import { generateToast, getUnifiedConfigs } from './util';
+import { parseErrorMsg } from './messageUtil';
+import { ResponseError } from './ResponseError';
+
+export interface RequestParams {
+    endpointUrl: string;
+    params?: Record<string, string | number>;
+    signal?: AbortSignal;
+    body?: BodyInit;
+    handleError: boolean;
+    callbackOnError?: (error: unknown) => void;
+}
+
+export function generateEndPointUrl(name: string) {
+    const unifiedConfigs = getUnifiedConfigs();
+    return `${unifiedConfigs.meta.restRoot}_${name}`;
+}
+
+const DEFAULT_PARAMS = { output_mode: 'json' };
+
+function createUrl(endpointUrl: string, params: Record<string, string | number>): URL {
+    const url = new URL(
+        createRESTURL(endpointUrl, { app, owner: 'nobody' }),
+        window.location.origin
+    );
+    Object.entries({ ...DEFAULT_PARAMS, ...params }).forEach(([key, value]) =>
+        url.searchParams.append(key, value.toString())
+    );
+    return url;
+}
+
+async function handleErrorResponse(response: Response): Promise<never> {
+    const errorData = await response.json();
+    const message = parseErrorMsg(errorData);
+    throw new ResponseError({ response, message });
+}
+
+async function fetchWithErrorHandling<TData>(
+    url: URL,
+    options: RequestInit,
+    handleError: boolean,
+    callbackOnError?: (error: unknown) => void
+): Promise<TData> {
+    const defaultInit = getDefaultFetchInit();
+
+    try {
+        const response = await fetch(url.toString(), {
+            ...defaultInit,
+            ...options,
+        });
+        if (!response.ok) {
+            await handleErrorResponse(response);
+        }
+        return await response.json();
+    } catch (error) {
+        if (handleError) {
+            const errorMsg = parseErrorMsg(error);
+            generateToast(errorMsg, 'error');
+            if (callbackOnError) {
+                callbackOnError(error);
+            }
+        }
+        throw error;
+    }
+}
```

Comment:
might be good to add some tests for those functions `generateEndPointUrl`, `handleErrorResponse` , `createUrl` ,  `fetchWithErrorHandling` 
as they are doing some modifications

Reply:
<img width="277" alt="image" src="https://github.com/user-attachments/assets/9e5d7080-44b4-4da4-aaab-ca468d9b0134">

Hm, 96% is a good coverage.
but when I looked inside

<img width="502" alt="image" src="https://github.com/user-attachments/assets/4be62a59-4128-448f-a69c-7ff5084025ba">
![image](https://github.com/user-attachments/assets/c6103e66-a690-4a79-a993-8ed942461ef4)


pretty suspicion. It should be covered by other tests. And I simply forgot to pass `callbackOnError` to `fetchWithErrorHandling`